source/Halibut/ServiceModel/PendingRequestQueueAsync.cs

```diff
@@ -83,27 +83,83 @@ async Task<PendingRequest> DequeueNextAsync(CancellationToken cancellationToken)
+            //// Option #1 - Task based - a bit leaky
+            //await Task.WhenAny(itemAddedToQueue.WaitAsync(cancellationToken), Task.Delay(pollingQueueWaitTimeout, cancellationToken));
+
+            // Option #1.1 - Task based with cancellation
+            using var cleanupCancellationTokenSource = new CancellationTokenSource();
+
+            using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cleanupCancellationTokenSource.Token);
+            await Task.WhenAny(itemAddedToQueue.WaitAsync(linkedCancellationTokenSource.Token), Task.Delay(pollingQueueWaitTimeout, linkedCancellationTokenSource.Token));
+
+            cleanupCancellationTokenSource.Cancel();
+
+            //// Option #2 - Cancellation Token Sources
+            //using var timeoutCancellationTokenSource = new CancellationTokenSource(pollingQueueWaitTimeout);
+            //using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token);
+
+            //try
+            //{
+            //    await itemAddedToQueue.WaitAsync(linkedCancellationTokenSource.Token);
+            //}
+            //catch (TaskCanceledException)
+            //{
+            //    cancellationToken.ThrowIfCancellationRequested();
+            //}
+
             itemAddedToQueue.Reset();
+
             return await TakeFirst(cancellationToken);
         }
 
         async Task<PendingRequest> TakeFirst(CancellationToken cancellationToken)
         {
-            using (await queueLock.LockAsync(cancellationToken))
+            //PendingRequest first = null;
+
+            //using (await queueLock.LockAsync(cancellationToken))
+            //{
+            //    if (!queue.TryDequeue(out first))
+            //    {
+            //        return null;
+            //    }
+                
+            //    if (queue.IsEmpty)
+            //    {
+            //        itemAddedToQueue.Reset();
+            //    }
+            //}
+
+            //if (first.IsDisposed)
+            //{
+            //    return await TakeFirst(cancellationToken);
+            //}
+
+            //return first;
+
+            PendingRequest first = null;
+
+            while (first == null)
             {
-                if (!queue.TryDequeue(out var first))
+                using (await queueLock.LockAsync(cancellationToken))
                 {
-                    return null;
+                    if (!queue.TryDequeue(out first))
+                    {
+                        return null;
+                    }
+
+                    if (queue.IsEmpty)
+                    {
+                        itemAddedToQueue.Reset();
+                    }
                 }
-                
-                if (queue.IsEmpty)
+
+                if (first.IsDisposed)
```

Comment:
Disposal of the PendingRequest is probably the biggest potential for bugs here and may make some of our tests that depend on a "dead" PendingRequest resulting in an instant null being returned to start failing as they will now wait the full `HalibutLimits.PollingQueueWaitTimeout` before null is returned